# Drosophila melanogaster

### Data preparation

Prior to running the experiments, follow the instructions in `data` and `annot`
folders to prepare input and annotation files

### GeneMark-ES

Run GeneMark-ES

```bash
cd ES
../../bin/gmes/gmes_petap.pl --verbose --seq \
    ../data/genome.fasta.masked --cores=8 --soft_mask auto --ES > log
cd ..
```

### GeneMark-ET

Run GeneMark-ET to compare protein results against RNA-Seq. RNA-Seq was sampled
and aligned by VARUS, see the `varus` folder for details.

```bash
mkdir ET; cd ET
../../bin/gmes/gmes_petap.pl --verbose  --seq ../data/genome.fasta.masked \
    --cores=8 --soft_mask auto --ET ../varus/varus.gff > log
cd ..
```

### ProtHint

Run ProtHint on all exclusion levels. Follow readme in `data` folder to
prepare protein data.

```bash
ls data | grep "\.fa$" | sed "s/\.fa//" | xargs -I{} bash -c '../bin/ProtHint/bin/prothint.py \
    data/genome.fasta.masked data/{}.fa --geneMarkGtf ES/genemark.gtf --workdir {} 2> logs/{}_log'
```

Make ProtHint accuracy table

```bash
../bin/create_prothint_accuracy_table.sh > accuracy_tables/prothint_accuracy.tsv
```

Generate start filtering table

```bash
../bin/create_start_filtering_table.sh family_excluded > accuracy_tables/start_filtering.tsv
```

To get a list of annotated genes which have at least a single hint in the hints file, use

```bash
../bin/get_hint_genes.sh ${level}_excluded/prothint.gff annot/annot.gtf genes_with_hints
```

### GeneMark-EP/EP+

Run GeneMark-EP/EP+ for all levels

```bash
../bin/EP_batch.sh species_excluded subgenus_excluded family_excluded order_excluded phylum_excluded
```

Make GeneMark-ES/EP/EP+ accuracy table

```bash
../bin/create_EP_accuracy_table.sh > accuracy_tables/es_ep_ep+_accuracy.tsv
```

Visualize EP+ results

```bash
../bin/visualize_EP+_results.sh annot/annot.gtf EP+_results_visualization cds 65 95 65 95
../bin/visualize_EP+_results.sh annot/annot.gtf EP+_results_visualization gene 37.5 77.5 37.5 77.5
../bin/visualize_EP+_results.sh annot/appris.gtf EP+_results_visualization/APPRIS cds 65 95 65 95
../bin/visualize_EP+_results.sh annot/appris.gtf EP+_results_visualization/APPRIS gene 35 75 35 75
```

### Annotation Statistics

Collect statistics about annotation

```bash
../bin/analyze_annot.sh annot/annot.gtf > accuracy_tables/annotation_stats.txt
```

### Extra run with no min exon score

Run ProtHint on family-excluded level without filtering out introns bordered by exons with
exon score < 25. This is done to generate a Sn-Sp curve for exon score. Figure with this curve
as well as scripts used to generate it are in `extra_runs/family_excluded_no_min_exon_score/graphs`
folder.

```bash
../bin/ProtHint/bin/prothint.py data/genome.fasta.masked data/family_excluded.fa \
    --geneMarkGtf ES/genemark.gtf --workdir extra_runs/family_excluded_no_min_exon_score \
    --minExonScore -1000 > logs/family_excluded_no_min_exon_score_log
```

Generate Sn-Sp figure

```bash
cd extra_runs/family_excluded_no_min_exon_score/graphs
./generateFigure.sh
cd ../../..
```

### Evaluating effects of hints in the plus mode

Run GeneMark-EP+ with introns only and with starts/stops only in the plus mode.

```bash
cd family_excluded/EP

# Introns only
mkdir plus_introns_only; cd plus_introns_only
grep Intron ../../evidence.gff > evidence.gff
../../../../bin/gmes/gmes_petap.pl --verbose --seq ../../../data/genome.fasta.masked \
    --cores=8 --soft_mask auto --EP ../../prothint.gff --evidence evidence.gff > log
cd ..

# Starts/stops only
mkdir plus_starts_stops_only; cd plus_starts_stops_only
grep -P "start_codon|stop_codon" ../../evidence.gff > evidence.gff
../../../../bin/gmes/gmes_petap.pl --verbose --seq ../../../data/genome.fasta.masked \
    --cores=8 --soft_mask auto --EP ../../prothint.gff --evidence evidence.gff > log
cd ../../..
```

Create a table with numbers of merged and split genes in ES, EP and EP+ with different
hints sets.

```bash
../bin/create_merging_splitting_table.sh family_excluded 10000 > accuracy_tables/merging_splitting.tsv
```

Accuracy table for EP+ with different hints sets.

```bash
../bin/create_plus_evidence_comparison_table.sh family_excluded > accuracy_tables/ep+_evidence_comparison.tsv
```

### ProtHint run with box kernel

Run ProtHint with a box instead of linear kernel for intron scoring to evaluate the
difference between kernels.

Since there is no interface for changing the kernel, this experiment requires a code adjustment of
`../bin/ProtHint/bin/run_spliced_alignment.pl` script.

In function `alignWithSpaln`, add an option`-k box` to `spaln_boundary_scorer` system call.
Then run ProtHint as usual:

```bash
../bin/ProtHint/bin/prothint.py data/genome.fasta.masked data/family_excluded.fa \
    --geneMarkGtf ES/genemark.gtf --workdir extra_runs/family_excluded_box_kernel \
    > logs/family_excluded_box_kernel_log
```

Generate Sn-Sp curves for IBA scores generated by box and linear kernels. Resulting figure is
located in `extra_runs/family_excluded_box_kernel/graphs`

```bash
cd extra_runs/family_excluded_box_kernel/graphs
./generateFigure.sh
cd ../../..
```

### Analysis of introns in conserved domains

The analysis of how many mapped ProtHint introns are located within regions coding for conserved protein domains
is documented in the `domains` folder.

### Second Iteration

Test the effect of a second iteration of ProtHint and EP+ using seeds from first
iteration of GeneMark-EP+ instead of GeneMark-ES.

```bash
../bin/ProtHint/bin/prothint.py data/genome.fasta.masked data/family_excluded.fa \
    --geneMarkGtf family_excluded/EP/plus/genemark.gtf --workdir extra_runs/family_excluded_iter_2 \
    > logs/family_excluded_iter_2_log
cd extra_runs/family_excluded_iter_2
mkdir -p EP/plus; cd EP/plus
../../../../../bin/gmes/gmes_petap.pl --verbose --seq \
    ../../../../data/genome.fasta.masked --cores=8 \
    --soft_mask auto --EP ../../prothint.gff --evidence ../../evidence.gff > log
cd ../../../..
```

### Experiment with conserved genes

Run GeneMark-EP+ with all High-Confidence (HC) introns vs HC introns which are found
in annotated genes fully supported by HC introns.

This experiment is done to see the positive effect of partially supported genes (only
some introns in a gene found by protein mapping).

```bash
cd extra_runs/ep_with_fully_supported_genes
./generate.sh
```

Result is stored in `extra_runs/ep_with_fully_supported_genes/visualization` folder.
